File: tests/property/python/areas/matrix.py
"""SageMath side of the ``matrix`` property-test area.

Matrix results are transported as nested integer lists so the transcript tests
the entries without depending on SageMath's pretty-printer.

Cases: tests/property/cases/matrix.cases.json
"""

from sage.all import ZZ, matrix


def _matrix(nrows, ncols, entries):
    return matrix(ZZ, nrows, ncols, entries)


def _rows(M):
    return [[ZZ(x) for x in row] for row in M.rows()]


def determinant_2x2(a, b, c, d):
    return _matrix(2, 2, [a, b, c, d]).determinant()


def determinant_3x3(a, b, c, d, e, f, g, h, i):
    return _matrix(3, 3, [a, b, c, d, e, f, g, h, i]).determinant()


def rank_2x3(a, b, c, d, e, f):
    return _matrix(2, 3, [a, b, c, d, e, f]).rank()


def rank_3x3(a, b, c, d, e, f, g, h, i):
    return _matrix(3, 3, [a, b, c, d, e, f, g, h, i]).rank()


def hnf_2x2(a, b, c, d):
    return _rows(_matrix(2, 2, [a, b, c, d]).hermite_form())


def hnf_3x3(a, b, c, d, e, f, g, h, i):
    return _rows(_matrix(3, 3, [a, b, c, d, e, f, g, h, i]).hermite_form())


def snf_2x2(a, b, c, d):
    return _rows(_matrix(2, 2, [a, b, c, d]).smith_form(transformation=False))


def snf_3x3(a, b, c, d, e, f, g, h, i):
    return _rows(
        _matrix(3, 3, [a, b, c, d, e, f, g, h, i]).smith_form(
            transformation=False
        )
    )


def lll_2x2(a, b, c, d):
    return _rows(_matrix(2, 2, [a, b, c, d]).LLL())


def lll_3x3(a, b, c, d, e, f, g, h, i):
    return _rows(_matrix(3, 3, [a, b, c, d, e, f, g, h, i]).LLL())


FUNCTIONS = {
    'determinant_2x2': determinant_2x2,
    'determinant_3x3': determinant_3x3,
    'rank_2x3': rank_2x3,
    'rank_3x3': rank_3x3,
    'hnf_2x2': hnf_2x2,
    'hnf_3x3': hnf_3x3,
    'snf_2x2': snf_2x2,
    'snf_3x3': snf_3x3,
    'lll_2x2': lll_2x2,
    'lll_3x3': lll_3x3,
}

File: tests/property/typescript/areas/matrix.ts
/**
 * sagemath-ts side of the `matrix` property-test area.
 *
 * Matrix results are transported as nested bigint arrays so the transcript
 * tests their entries without depending on either implementation's printer.
 *
 * Cases: tests/property/cases/matrix.cases.json
 * SageMath counterpart: tests/property/python/areas/matrix.py
 */

import {
  type IntegerMatrix,
  IntegerMatrixFromEntries,
  LLL,
  hermite_normal_form,
  smith_form_integer,
} from '../../../../packages/sagemath-ts/src/matrix/index.js';

function matrixRows(matrix: IntegerMatrix): bigint[][] {
  const rows: bigint[][] = [];
  for (let i = 0; i < matrix.nrows; i++) {
    const row: bigint[] = [];
    for (let j = 0; j < matrix.ncols; j++) {
      row.push(matrix.get(i, j).value);
    }
    rows.push(row);
  }
  return rows;
}

function matrix2(a: bigint, b: bigint, c: bigint, d: bigint): IntegerMatrix {
  return IntegerMatrixFromEntries([
    [a, b],
    [c, d],
  ]);
}

function matrix3(
  a: bigint,
  b: bigint,
  c: bigint,
  d: bigint,
  e: bigint,
  f: bigint,
  g: bigint,
  h: bigint,
  i: bigint
): IntegerMatrix {
  return IntegerMatrixFromEntries([
    [a, b, c],
    [d, e, f],
    [g, h, i],
  ]);
}

export const functions = {
  determinant_2x2: (a: bigint, b: bigint, c: bigint, d: bigint): bigint =>
    matrix2(a, b, c, d).determinant().value,

  determinant_3x3: (
    a: bigint,
    b: bigint,
    c: bigint,
    d: bigint,
    e: bigint,
    f: bigint,
    g: bigint,
    h: bigint,
    i: bigint
  ): bigint => matrix3(a, b, c, d, e, f, g, h, i).determinant().value,

  rank_2x3: (a: bigint, b: bigint, c: bigint, d: bigint, e: bigint, f: bigint): number =>
    IntegerMatrixFromEntries([
      [a, b, c],
      [d, e, f],
    ]).rank(),

  rank_3x3: (
    a: bigint,
    b: bigint,
    c: bigint,
    d: bigint,
    e: bigint,
    f: bigint,
    g: bigint,
    h: bigint,
    i: bigint
  ): number => matrix3(a, b, c, d, e, f, g, h, i).rank(),

  hnf_2x2: (a: bigint, b: bigint, c: bigint, d: bigint): bigint[][] =>
    matrixRows(hermite_normal_form(matrix2(a, b, c, d)) as IntegerMatrix),

  hnf_3x3: (
    a: bigint,
    b: bigint,
    c: bigint,
    d: bigint,
    e: bigint,
    f: bigint,
    g: bigint,
    h: bigint,
    i: bigint
  ): bigint[][] =>
    matrixRows(hermite_normal_form(matrix3(a, b, c, d, e, f, g, h, i)) as IntegerMatrix),

  snf_2x2: (a: bigint, b: bigint, c: bigint, d: bigint): bigint[][] =>
    matrixRows(smith_form_integer(matrix2(a, b, c, d)) as IntegerMatrix),

  snf_3x3: (
    a: bigint,
    b: bigint,
    c: bigint,
    d: bigint,
    e: bigint,
    f: bigint,
    g: bigint,
    h: bigint,
    i: bigint
  ): bigint[][] =>
    matrixRows(smith_form_integer(matrix3(a, b, c, d, e, f, g, h, i)) as IntegerMatrix),

  lll_2x2: (a: bigint, b: bigint, c: bigint, d: bigint): bigint[][] =>
    matrixRows(LLL(matrix2(a, b, c, d)) as IntegerMatrix),

  lll_3x3: (
    a: bigint,
    b: bigint,
    c: bigint,
    d: bigint,
    e: bigint,
    f: bigint,
    g: bigint,
    h: bigint,
    i: bigint
  ): bigint[][] => matrixRows(LLL(matrix3(a, b, c, d, e, f, g, h, i)) as IntegerMatrix),
};
